File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/brasil/BrasilSemardriveCrawler.java
package br.com.lett.crawlernode.crawlers.ranking.keywords.brasil;

import br.com.lett.crawlernode.core.fetcher.ProxyCollection;
import br.com.lett.crawlernode.core.fetcher.methods.JavanetDataFetcher;
import br.com.lett.crawlernode.core.fetcher.models.Request;
import br.com.lett.crawlernode.core.models.RankingProduct;
import br.com.lett.crawlernode.core.models.RankingProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords;
import br.com.lett.crawlernode.exceptions.MalformedProductException;
import br.com.lett.crawlernode.util.CrawlerUtils;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class BrasilSemardriveCrawler extends CrawlerRankingKeywords {

   public BrasilSemardriveCrawler(Session session) {
      super(session);
   }

   protected String getProductsList() {
      String url = "https://www.semarentrega.com.br/ccstoreui/v1/search?suppressResults=false&searchType=simple&No=1&Nrpp=24&Ntt=" + keywordEncoded + "&page=" + (currentPage - 1);

      Request request = Request.RequestBuilder.create()
         .setUrl(url)
         .setProxyservice(Arrays.asList(ProxyCollection.NETNUT_RESIDENTIAL_BR))
         .build();

      String response = dataFetcher.get(session, request).getBody();

      return response;

   }

   @Override
   protected void extractProductsFromCurrentPage() throws UnsupportedEncodingException, MalformedProductException {
      String response = getProductsList();
      JSONObject reponseJson = CrawlerUtils.stringToJson(response);
      this.totalProducts = reponseJson.optJSONObject("resultsList").optInt("totalNumRecs");
      if (totalProducts > 0) {
         JSONArray productsList = reponseJson.getJSONObject("resultsList").getJSONArray("records");
         for (Object arrayOfArrays : productsList) {
            JSONObject jsonInfo = (JSONObject) arrayOfArrays;
            JSONArray records = jsonInfo.getJSONArray("records");
            String productUrl = "https://www.semarentrega.com.br" + records.optQuery("/0/attributes/product.route/0").toString();
            String internalId = jsonInfo.optQuery("/attributes/product.repositoryId/0").toString();
            String name = records.optQuery("/0/attributes/sku.displayName/0").toString();
            String imgUrl = "https://www.semarentrega.com.br" + records.optQuery("/0/attributes/product.primaryFullImageURL/0").toString();

            String availability = records.optQuery("/0/attributes/sku.availabilityStatus/0").toString();
            boolean isAvailable = availability.isEmpty() || !availability.contains("OUTOFSTOCK");
            Integer price = null;
            if(isAvailable == true){
               String priceText = (String) records.optQuery("/0/attributes/product.listPrice/0");
               if(priceText != null && !priceText.isEmpty()){
                  Double priceDouble = Double.parseDouble(priceText) * 100;
                  price = priceDouble.intValue();
               }
            }




            RankingProduct productRanking = RankingProductBuilder.create()
               .setUrl(productUrl)
               .setInternalId(internalId)
               .setName(name)
               .setImageUrl(imgUrl)
               .setPriceInCents(price)
               .setAvailability(isAvailable)
               .build();

            saveDataProduct(productRanking);

         }
      } else {
         this.result = false;
         this.log("Keyword sem resultado!");
      }

   }
}

File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/brasil/BrasilSemardriveCrawler.kt
package br.com.lett.crawlernode.crawlers.ranking.keywords.brasil

import br.com.lett.crawlernode.core.fetcher.FetchMode
import br.com.lett.crawlernode.core.fetcher.methods.FetcherDataFetcher
import br.com.lett.crawlernode.core.fetcher.models.Request
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.util.HashMap

/**
 * Date: 22/10/20
 *
 * @author Fellype Layunne
 *
 */
abstract class BrasilSemardriveCrawler(session: Session) : CrawlerRankingKeywords(session) {

   companion object {
      const val HOME_PAGE = "https://drive.gruposemar.com.br/products"
   }

   init {
      fetchMode = FetchMode.APACHE
   }

   abstract fun getZipCode(): String

   fun handleCookies() {
      val url = "https://drive.gruposemar.com.br/current_stock"

      val headers: MutableMap<String, String> = HashMap()
      headers["content-type"] = "application/x-www-form-urlencoded; charset=UTF-8"
      headers["origin"] = "https://drive.gruposemar.com.br"
      headers["accept"] = "*/*;q=0.5, text/javascript, application/javascript, application/ecmascript, application/x-ecmascript"

      val payload = "utf8=%E2%9C%93" +
         "&_method=put" +
         "&order%5Bshipping_mode%5D=delivery" +
         "&order%5Bship_address_attributes%5D%5Btemporary%5D=true" +
         "&order%5Bship_address_attributes%5D%5Bzipcode%5D=${getZipCode()}"

      val response = FetcherDataFetcher().post(
         session, Request.RequestBuilder.create()
         .setUrl(url)
         .setPayload(payload)
         .setHeaders(headers)
         .build()
      )
      cookies = response.cookies
   }

   fun fetchDocument(): Document {
      val url = "${HOME_PAGE}?keywords=${keywordEncoded}&page=${currentPage}&utf8=%E2%9C%93"

      val token = cookies.first {
         it.name == "guest_token"
      }?.value ?: ""

      val headers: MutableMap<String, String> = HashMap()

      headers["cookie"] = "guest_token=${token};"
      headers["authority"] = "drive.gruposemar.com.br"
      headers["accept"] = "text/html, application/xhtml+xml"
      headers["user-agent"] = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36"

      val request = Request.RequestBuilder.create()
         .setUrl(url)
         .setHeaders(headers)
         .build()

      val response = dataFetcher.get(session, request)

      return Jsoup.parse(response?.body ?: "")
   }

   override fun extractProductsFromCurrentPage() {

      if (currentPage == 1) {
         handleCookies()
      }

      currentDoc = fetchDocument()

      val products = currentDoc.select("#product-results .product")

      for (product in products) {

         val internalId = product.selectFirst(".buttons input[name=variant_id]")?.attr("value")

         val path = product.selectFirst(".text a")?.attr("href") ?: ""

         val productUrl = "$HOME_PAGE$path"

         saveDataProduct(internalId, internalId, productUrl)
         log(">>> productId: $internalId || url: $productUrl || name: ${product.selectFirst(".text .product-title")?.text() ?: ""}")
      }
   }

   override fun hasNextPage(): Boolean {
      return currentDoc.selectFirst("#products-paginator a") != null
   }
}
